unsafe/src/main/java/org/apache/spark/unsafe/types/Interval.java

```diff
@@ -0,0 +1,330 @@
+    unitRegex("week") + unitRegex("day") + unitRegex("hour") + unitRegex("minute") +
+    unitRegex("second") + unitRegex("millisecond") + unitRegex("microsecond");
+
+  private static Pattern calendarIntervalPattern =
+    Pattern.compile("interval" + unitRegex("year") + unitRegex("month") + dayTimePatternString);
+
+  private static Pattern timeIntervalPattern = Pattern.compile("interval" + dayTimePatternString);
+
+  private static Pattern yearMonthPattern =
+    Pattern.compile("^(?:['|\"])?([+|-])?(\\d+)-(\\d+)(?:['|\"])?$");
+
+  private static Pattern dayTimePattern =
+    Pattern.compile("^(?:['|\"])?([+|-])?(\\d+) (\\d+):(\\d+):(\\d+)(\\.(\\d+))?(?:['|\"])?$");
+
+  private static Pattern quoteTrimPattern = Pattern.compile("^(?:['|\"])?(.*?)(?:['|\"])?$");
+
+  /**
+   * Parse interval term string like "interval 5 years 3 months 1 weeks 2 days 5 seconds" to its
+   * internal representation [[CalendarInterval]].
+   */
+  public static CalendarInterval stringToCalendarInterval(String s)
+      throws IllegalArgumentException {
+    if (s == null) {
+      return null;
+    }
+    s = s.trim();
+    Matcher m = calendarIntervalPattern.matcher(s);
+    if (!m.matches() || s.equals("interval")) {
+      throw new IllegalArgumentException("Invalid interval String " + s);
+    } else {
+      long months = toLong(m.group(1)) * 12 + toLong(m.group(2));
+      long microseconds = toLong(m.group(3)) * MICROS_PER_WEEK;
+      microseconds += toLong(m.group(4)) * MICROS_PER_DAY;
+      microseconds += toLong(m.group(5)) * MICROS_PER_HOUR;
+      microseconds += toLong(m.group(6)) * MICROS_PER_MINUTE;
+      microseconds += toLong(m.group(7)) * MICROS_PER_SECOND;
+      microseconds += toLong(m.group(8)) * MICROS_PER_MILLI;
+      microseconds += toLong(m.group(9));
+      return new CalendarInterval((int) months, microseconds);
+    }
+  }
+
+  /**
+   * Parse interval term string like "interval 1 weeks 2 days 5 seconds" to its
+   * internal representation - a Long value for number of microseconds.
+   */
+  public static long stringToTimeInterval(String s) throws IllegalArgumentException {
+    if (s == null) {
+      return 0;
+    }
+    s = s.trim();
+    Matcher m = timeIntervalPattern.matcher(s);
+    if (!m.matches() || s.equals("interval")) {
+      throw new IllegalArgumentException("Invalid interval String " + s);
+    } else {
+      long microseconds = toLong(m.group(1)) * MICROS_PER_WEEK;
+      microseconds += toLong(m.group(2)) * MICROS_PER_DAY;
+      microseconds += toLong(m.group(3)) * MICROS_PER_HOUR;
+      microseconds += toLong(m.group(4)) * MICROS_PER_MINUTE;
+      microseconds += toLong(m.group(5)) * MICROS_PER_SECOND;
+      microseconds += toLong(m.group(6)) * MICROS_PER_MILLI;
+      microseconds += toLong(m.group(7));
+      return microseconds;
+    }
+  }
+
+  public static long calendarIntervalToTimeInterval(CalendarInterval calendarInterval) {
+    return calendarInterval.microseconds;
```

Comment:
Hi @cloud-fan, sorry for my late reply, I agree with your idea that the conversion should be under some context instead of just cut off the month part. For the cast from `CalendarInterval` to `TimeInterval`, I think it's more reasonable to just use the truncated one (as the default context)? Since there aren't any context available.